### ISSUE TITLE
Issue when loading multiple hooks with this module

### ISSUE THREAD

**Issue by Pierre-Gilles** [jaumard/sails-util-mvcsloader#1]
Hi !

I'm working on a big sails.js project, and your module is exactly what I'm looking for! 

I have multiple hooks in my application, and all hooks needs to inject their models, controllers, policies,   into the main app.
With one hooks it works really fine, but when I've got two hooks in my 'api/hooks' folder, I've got a big error : 
```
error: Failed to reinitialize ORM.
error: AdapterError: Connection is already registered
    at Object.registerConnection (/home/pi/Gladys/node_modules/sails-mysql/lib/connections/register.js:27:58)
    at /home/pi/Gladys/node_modules/sails/node_modules/waterline/lib/waterline.js:203:29
    at /home/pi/Gladys/node_modules/async/lib/async.js:125:13
    at Array.forEach (native)
    at _each (/home/pi/Gladys/node_modules/async/lib/async.js:46:24)
    at Object.async.each (/home/pi/Gladys/node_modules/async/lib/async.js:124:9)
    at Array.async.auto.registerConnections [as 1] (/home/pi/Gladys/node_modules/sails/node_modules/waterline/lib/waterline.js:176:13)
    at /home/pi/Gladys/node_modules/async/lib/async.js:487:38
    at Array.forEach (native)
    at _each (/home/pi/Gladys/node_modules/async/lib/async.js:46:24)
    at Object.async.auto (/home/pi/Gladys/node_modules/async/lib/async.js:458:9)
    at Waterline.initialize (/home/pi/Gladys/node_modules/sails/node_modules/waterline/lib/waterline.js:121:9)
    at buildORM (/home/pi/Gladys/node_modules/sails/lib/hooks/orm/build-orm.js:52:15)
    at Array.async.auto.instantiatedCollections [as 1] (/home/pi/Gladys/node_modules/sails/lib/hooks/orm/index.js:203:11)
    at listener (/home/pi/Gladys/node_modules/async/lib/async.js:493:46)
    at /home/pi/Gladys/node_modules/async/lib/async.js:444:17 { [AdapterError: Connection is already registered]
  name: 'AdapterError',
  message: 'Connection is already registered' }

/home/pi/Gladys/node_modules/sails/lib/hooks/orm/index.js:221
            throw new Error(err);
```

Any idea how to solve this problem ?

Second issues, in your example hooks, is it really : 
```
var loader = require("sails-hook-hookloader")(sails);
```
and not : 
```
var loader = require("sails-util-mvcsloader")(sails);
```
? 

Thanks for your answer !

( in spite of these little errors, your module rocks !)

**Comment by jaumard** [comment]
Hi, 

I never test with more than 2 hooks ! :) The problem is when the first hook finish to load models I reload the ORM (to take care of new models). I suppose it can't be reloaded multiple times if it's already reloading (cause hook are loaded asynchronously).

If you want a quick fix you can test to comment/remove sails.hooks.orm.reload(); on libs/models.js (line 47 or arround) in this module and do it on your config/bootstrap.js.

Let me know if it's working.
I will search a proper way to fix this. 

Thanks :)

**Comment by Pierre-Gilles** [comment]
Ok it work fine with your quick fix :)

It would be great to be able to do this natively in the module, for example by disabling the hook reload with a parameter in your loader.injectAll function, or something else I don't know!

Thanks for your quick answer,

**Comment by Pierre-Gilles** [comment]
Do you know if there is a way to know when sails.hooks.orm.reload(); has finished ? ( seems to don't accept callbacks )

Because I was doing other things in my bootstrap function, like cleaning tables in the database, and it don't work anymore because the sails.hooks.orm.reload(); is in process..

**Comment by jaumard** [comment]
Great !! Happy to ear this,

I don't have to test right now but if you have time please test this : 

Remove under index.js of this module the sails.on('hook:orm:loaded', function (){}); (around line 142) but the content of the callback have to stay. 
Let the ORM reload commented but don't do it on your bootstrap.js.

I think like this it's working ! 
I was looking for a isReloading function on hook ORM  but don't find one. But with this fix no need to reload it :)

**Comment by Pierre-Gilles** [comment]
Perfect that work fine ! :) 

Thank you for answers, I'm waiting for the module update to not have to change it manually each time ;) 

Do you know when you'll be supporting views and assets ? I'm doing it by hand for the moment, by simply copy/pasting assets in the assets folder, but that's not a good way to work I think.

**Comment by jaumard** [comment]
If it's not working you can do this on your bootstrap : 

      sails.hooks.orm.reload();
      sails.on("hook:orm:reloaded", function ()
      {
        console.log("test");
        cb();
      });

**Comment by jaumard** [comment]
Ho surprise that's working without an ORM reload ^^ which database did you use ? Did you use schema : true on your config/models.js ? 
I will make more tests and I will update the module ! :) 

Thanks for giving me feedbacks :)

**Comment by Pierre-Gilles** [comment]
In fact I haven't had the time to check properly if all has been injected into the main app! I'll check tonight :)
But the server does not crash anymore, and the sails.log.info of your module are showing beautiful message so it's nicer than before !

I'm using sails-mysql ! I don't remember if I use schema true or not, I'm on the road now I'll check tonight.

**Comment by jaumard** [comment]
Great let me know, I will check soon too I think.
For views and assets it's quite hard to implement cause views are compiled and assets are sync on .tmp folder, all this by grunt tasks :/ for now it's simpler to copy/paste it with a postinstall script like I do on this module https://github.com/jaumard/sails-hook-passport

**Comment by jaumard** [comment]
I just test and it's not working if I remove hook:orm event, there no error but the orm doesn't create the models... Can you confirm when you have time please :) 
I think I will put a boolean for reloading ORM or not for now.

**Comment by Pierre-Gilles** [comment]
Same for me ! No error but no models created.

I think a boolean is fine for now :)

**Comment by jaumard** [comment]
Ok I just put 0.1.6 version online please try it and let me know if it's ok :) 
I update the README file https://github.com/jaumard/sails-util-mvcsloader#troubles

**Comment by Pierre-Gilles** [comment]
I've tried it but it seems that it's not working :/

The server does not crash, but the models are not created in my database ( sails-mysql ).

I've created the config file in config/mvcsloader.js, and put s"ails.hooks.orm.reload();" in the bootstrap file but it don't work

**Comment by jaumard** [comment]
Strange it's working for me on a new project, maybe some cache or something like this ? Can you test on a new clean project ?

Thanks

**Comment by jaumard** [comment]
Ok just found a way to fix this, I put a new version 0.1.7 tell me if it's working. I remove the boolean is not needed anymore :) the method to remove event is working it was just an async problem.

**Comment by Pierre-Gilles** [comment]
Perfect I took time to test it today and it work perfectly !

I close the issue, thank you for your time :)
